File: objc/Tools/TtioVerify.m
/*
 * TtioVerify — minimal CLI that opens a .tio file and prints a JSON
 * summary suitable for the Python cross-compat test (M16). The output is
 * intentionally a flat, stable shape so the Python harness can parse it
 * without depending on a full ObjC reader.
 *
 * Build with gnustep-make (see GNUmakefile).
 * Usage: TtioVerify <path-to-.tio>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

#import <Foundation/Foundation.h>

#import "Dataset/TTIOSpectralDataset.h"
#import "Dataset/TTIOIdentification.h"
#import "Dataset/TTIOQuantification.h"
#import "Dataset/TTIOProvenanceRecord.h"
#import "Run/TTIOAcquisitionRun.h"
#import "Run/TTIOSpectrumIndex.h"
#import "Genomics/TTIOGenomicRun.h"

static NSString *JSONEscape(NSString *s)
{
    if (!s) return @"\"\"";
    NSMutableString *out = [NSMutableString stringWithString:@"\""];
    for (NSUInteger i = 0; i < s.length; i++) {
        unichar c = [s characterAtIndex:i];
        switch (c) {
            case '"':  [out appendString:@"\\\""]; break;
            case '\\': [out appendString:@"\\\\"]; break;
            case '\n': [out appendString:@"\\n"];  break;
            case '\r': [out appendString:@"\\r"];  break;
            case '\t': [out appendString:@"\\t"];  break;
            default:
                if (c < 0x20) [out appendFormat:@"\\u%04x", c];
                else          [out appendFormat:@"%C", c];
        }
    }
    [out appendString:@"\""];
    return out;
}

int main(int argc, const char *argv[])
{
    @autoreleasepool {
        if (argc < 2) {
            fprintf(stderr, "usage: TtioVerify <path-to-.tio>\n");
            return 2;
        }
        NSString *path = [NSString stringWithUTF8String:argv[1]];

        NSError *err = nil;
        TTIOSpectralDataset *ds =
            [TTIOSpectralDataset readFromFilePath:path error:&err];
        if (!ds) {
            fprintf(stderr, "TtioVerify: failed to open %s: %s\n",
                    argv[1], err.localizedDescription.UTF8String ?: "");
            return 1;
        }

        NSMutableString *out = [NSMutableString stringWithString:@"{"];
        [out appendFormat:@"\"title\":%@,", JSONEscape(ds.title)];
        [out appendFormat:@"\"isa_investigation_id\":%@,",
            JSONEscape(ds.isaInvestigationId)];

        // MS runs
        [out appendString:@"\"ms_runs\":{"];
        BOOL firstRun = YES;
        NSArray *runNames = [[ds.msRuns allKeys]
            sortedArrayUsingSelector:@selector(compare:)];
        for (NSString *rname in runNames) {
            TTIOAcquisitionRun *run = ds.msRuns[rname];
            if (!firstRun) [out appendString:@","];
            firstRun = NO;
            [out appendFormat:@"%@:{\"spectrum_count\":%lu}",
                JSONEscape(rname), (unsigned long)run.spectrumIndex.count];
        }
        [out appendString:@"},"];

        // Genomic runs (M82.4 cross-language conformance matrix).
        [out appendString:@"\"genomic_runs\":{"];
        BOOL firstGRun = YES;
        NSArray *gRunNames = [[ds.genomicRuns allKeys]
            sortedArrayUsingSelector:@selector(compare:)];
        for (NSString *gname in gRunNames) {
            TTIOGenomicRun *gr = ds.genomicRuns[gname];
            if (!firstGRun) [out appendString:@","];
            firstGRun = NO;
            [out appendFormat:@"%@:{\"read_count\":%lu,"
                              @"\"reference_uri\":%@,"
                              @"\"platform\":%@,"
                              @"\"sample_name\":%@}",
                JSONEscape(gname), (unsigned long)gr.readCount,
                JSONEscape(gr.referenceUri),
                JSONEscape(gr.platform),
                JSONEscape(gr.sampleName)];
        }
        [out appendString:@"},"];

        [out appendFormat:@"\"identification_count\":%lu,",
            (unsigned long)ds.identifications.count];
        [out appendFormat:@"\"quantification_count\":%lu,",
            (unsigned long)ds.quantifications.count];
        [out appendFormat:@"\"provenance_count\":%lu",
            (unsigned long)ds.provenanceRecords.count];
        [out appendString:@"}"];

        puts([out UTF8String]);
        return 0;
    }
}

File: java/src/main/java/global/thalion/ttio/tools/TtioVerify.java
/*
 * TTI-O Java Implementation
 * Copyright (C) 2026 DTW-Thalion
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio.tools;

import global.thalion.ttio.AcquisitionRun;
import global.thalion.ttio.SpectralDataset;
import global.thalion.ttio.genomics.GenomicRun;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Java equivalent of the Objective-C {@code TtioVerify} CLI. Reads
 * a .tio file and prints a flat JSON summary that matches the ObjC
 * tool's output byte-for-byte (modulo whitespace), so a Python
 * cross-language harness can spawn both subprocesses and assert
 * structural equivalence.
 *
 * <p>Output schema (stable):</p>
 * <pre>
 * {
 *   "title": "...",
 *   "isa_investigation_id": "...",
 *   "ms_runs": {"run_0001": {"spectrum_count": 5}, ...},
 *   "genomic_runs": {"genomic_0001": {"read_count": 100,
 *                                      "reference_uri": "...",
 *                                      "platform": "...",
 *                                      "sample_name": "..."}, ...},
 *   "identification_count": N,
 *   "quantification_count": N,
 *   "provenance_count": N
 * }
 * </pre>
 *
 * <p>The {@code genomic_runs} block was added in M82.4 to drive the
 * cross-language conformance matrix; pre-M82 datasets emit it as an
 * empty object.</p>
 *
 * <p>Usage: {@code mvn exec:java -Dexec.mainClass=global.thalion.ttio.tools.TtioVerify -Dexec.args="path/to.tio"}</p>
 *
 * @since 0.9
 */
public final class TtioVerify {

    private TtioVerify() {}

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("usage: TtioVerify <path-to-.tio>");
            System.exit(2);
        }
        try (SpectralDataset ds = SpectralDataset.open(args[0])) {
            StringBuilder out = new StringBuilder("{");
            out.append("\"title\":").append(jsonEscape(ds.title())).append(",");
            out.append("\"isa_investigation_id\":")
               .append(jsonEscape(ds.isaInvestigationId())).append(",");

            out.append("\"ms_runs\":{");
            List<String> names = new ArrayList<>(ds.msRuns().keySet());
            Collections.sort(names);
            for (int i = 0; i < names.size(); i++) {
                String n = names.get(i);
                AcquisitionRun run = ds.msRuns().get(n);
                if (i > 0) out.append(",");
                out.append(jsonEscape(n))
                   .append(":{\"spectrum_count\":")
                   .append(run.spectrumCount())
                   .append("}");
            }
            out.append("},");

            out.append("\"genomic_runs\":{");
            List<String> gnames = new ArrayList<>(ds.genomicRuns().keySet());
            Collections.sort(gnames);
            for (int i = 0; i < gnames.size(); i++) {
                String n = gnames.get(i);
                GenomicRun gr = ds.genomicRuns().get(n);
                if (i > 0) out.append(",");
                out.append(jsonEscape(n))
                   .append(":{\"read_count\":").append(gr.readCount())
                   .append(",\"reference_uri\":").append(jsonEscape(gr.referenceUri()))
                   .append(",\"platform\":").append(jsonEscape(gr.platform()))
                   .append(",\"sample_name\":").append(jsonEscape(gr.sampleName()))
                   .append("}");
            }
            out.append("},");

            out.append("\"identification_count\":")
               .append(ds.identifications().size()).append(",");
            out.append("\"quantification_count\":")
               .append(ds.quantifications().size()).append(",");
            out.append("\"provenance_count\":")
               .append(ds.provenanceRecords().size());
            out.append("}");
            System.out.println(out);
        } catch (Exception e) {
            System.err.println("TtioVerify: failed to open " + args[0] + ": " + e.getMessage());
            System.exit(1);
        }
    }

    private static String jsonEscape(String s) {
        if (s == null) return "\"\"";
        StringBuilder sb = new StringBuilder("\"");
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\n': sb.append("\\n");  break;
                case '\r': sb.append("\\r");  break;
                case '\t': sb.append("\\t");  break;
                default:
                    if (c < 0x20) sb.append(String.format("\\u%04x", (int) c));
                    else          sb.append(c);
            }
        }
        sb.append('"');
        return sb.toString();
    }
}
